anvil/distributions.py

```diff
@@ -120,16 +299,17 @@ class PhiFourAction(nn.Module):
 
     def __init__(self, m_sq, lam, geometry, use_arxiv_version=False):
         super(PhiFourAction, self).__init__()
-        self.geometry = geometry
-        self.shift = self.geometry.get_shift()
+        self.shift = geometry.get_shift()
         self.lam = lam
         self.m_sq = m_sq
-        self.length = self.geometry.length
+        self.lattice_size = geometry.length ** 2
         if use_arxiv_version:
             self.version_factor = 2
         else:
             self.version_factor = 1
 
+        self.log_density = self.forward
```

Comment:
Are we actually using the forward method anywhere? I don't know if there are weird side effects of doing this because pytorch messes around with the forward method so that gradients are tracked.

I wonder if we can just drop the forward method all together and just have a log_density then I think (but we should check) that we could not inherit from the nn.Module class but I might be wrong

if not then I think I would prefer if we had something like `Target(nn.Module)` which had `def log_density(self, states): return self.forward(states)` and then inheritted from that instead

Reply:
> I don't know if there are weird side effects of doing this because pytorch messes around with the forward method so that gradients are tracked.

Thanks for flagging this up - I had not given this anywhere near enough thought...

I believe the only really important thing we get from subclassing `nn.Module` is that the optimizer is automatically made aware of any tunable parameters. I think we should definitely still subclass `nn.Module` for our normalising flow models because they are what we pass to the optimizer so they should contain the list of parameters.
Also, they're kind of "part of the network" and it's neat, plus we may want to add tunable parameters in future (e.g. a scaling factor inside a tanh).

However, when we do `current_loss.backward()` this really shouldn't complain if some of the operations in the tree aren't collected into a module forward method - it should (I think) be fine as long as every intermediate tensor has `requires_grad = True`.

So I suppose it makes sense to just have all of the base and target distributions as regular classes. That way everything can have a `log_density` method so it can be used as a target, and anything which can act as a base distribution can additionally have a method which does
```python
z = generate_random_sample(sample_size)
return z, self.log_density(z)
```